SeQuant/core/space.hpp

```diff
@@ -417,109 +314,446 @@ class IndexSpace {
+// check to see if the resulting space is actually registered.
+  IndexSpace intersection(const IndexSpace &space1,
+                                        const IndexSpace &space2,
+                                        const IndexSpace &space3) {
+    if( space1 == space2 && space1 == space3){
+      return space1;
+    }
+    else {
+      bool same_qns = ((space1.qns() == space2.qns()) && (space1.qns() == space3.qns()));
+      if(!same_qns){throw std::invalid_argument("asking for the intersection of spaces with incompatible quantum number attributes.");}
+      auto intersection_attr = space1.type().intersection(space2.type()).intersection(space3.type());
+      IndexSpace intersection_space = find_IndexSpace({intersection_attr,space1.qns()});
+      if(intersection_space == nulltype && intersection_attr != 0){
+        throw std::invalid_argument("The resulting space is not registered in this context. Add this "
+            "space to the registry with a label to use it.");
+      }
+      else{return intersection_space;}
+    }
+  }
+
+  //return the resulting space corresponding to a bitwise unIon between two spaces.
+  // check to see if the resulting space is actually registered.
+  IndexSpace unIon(const IndexSpace &space1,
+                                 const IndexSpace &space2) {
+    if( space1 == space2){
+      return space1;
+    }
+    else{
+      bool same_qns = space1.qns() == space2.qns();
+      if(!same_qns){throw std::invalid_argument("asking for the intersection of spaces with incompatible quantum number attributes.");}
+      auto unIontype = space1.type().unIon(space2.type());
+      IndexSpace unIonSpace = find_IndexSpace({unIontype,space1.qns()});
+      if(unIonSpace == nulltype){
+        throw std::invalid_argument("The resulting space is not registered in this context. Add this "
+            "space to the registry with a label to use it.");
+      }
+      else{return unIonSpace;}
+    }
+  }
+
+  //return the resulting spaces corresponding to a bitwise excluded or between two spaces.
+  // check to see if the resulting space is actually registered.
+  std::vector<IndexSpace> non_overlapping_spaces(const IndexSpace &space1, const IndexSpace &space2)const {
+    auto attributes = space1.attr().excluded_spaces(space2.attr());
+    std::vector<IndexSpace> result;
+    for(int i = 0; i < attributes.size(); i++){
+      auto excluded_space = find_IndexSpace(attributes[i]);
+      if(excluded_space == nulltype){
+        throw std::invalid_argument("The resulting space is not registered in this context. Add this "
+            "space to the registry with a label to use it.");
+      }
+      result.push_back(excluded_space);
+    }
+    return result;
+  }
+
+// do two spaces have non_overlapping spaces
+bool has_non_overlapping_spaces(const IndexSpace &space1, const IndexSpace &space2)const{
+if(space1.type().exclusionary_or(space2.type()).to_int32() == 0) {return false;}
+else{ return true;}
```

Comment:
```suggestion
return space1.type().exclusionary_or(space2.type()).to_int32() != 0;
```
No need for any branching when all you want to do is return a bool